src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs

```diff
@@ -154,34 +154,39 @@ public static MemoryMappedFile CreateFromFile(string path, FileMode mode, string
                 FileMode.CreateNew => false,
                 _ => File.Exists(path)
             };
-            FileStream fileStream = new FileStream(path, mode, GetFileAccess(access), FileShare.Read, 0x1000, FileOptions.None);
+            SafeFileHandle fileHandle = File.OpenHandle(path, mode, GetFileAccess(access), FileShare.Read, FileOptions.None);
+            long fileSize = mode switch
+            {
+                FileMode.CreateNew or FileMode.Create => 0, // the file is brand new
+                _ => RandomAccess.GetLength(fileHandle)
```

Comment:
If GetLength throws, the file handle won't be Dispose'd.

Reply:
Also, doesn't this throw for non-regular files?  Is the exception that occurs in such a case the same as it was before?

Reply:
> If GetLength throws, the file handle won't be Dispose'd.

that is correct, I am going to add a try-catch block

> Also, doesn't this throw for non-regular files?

both `FileStream.Length` and `RandomAccess.GetLength` throw for non-seekable files. 

the special character device files report that they are seekable and of a size equal zero

Reply:
Do we have tests that validate the exception in this case, both with capacity == 0 and capacity != 0?

Reply:
> Do we have tests that validate the exception in this case, both with capacity == 0 and capacity != 0?


We don't. For both cases, the current implementation would throw for non-seekable file (without closing the handle).

For `capacity == 0` here when accessing `fileStream.Length`:

https://github.com/dotnet/runtime/blob/8d8ef6dac58acf2ab944289c69dd46ee9e12d717/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs#L159

For `capacity != 0` here when accessing `fileStream.Length`:

https://github.com/dotnet/runtime/blob/8d8ef6dac58acf2ab944289c69dd46ee9e12d717/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Windows.cs#L16-L22

Reply:
@stephentoub I've sent a PR with tests https://github.com/dotnet/runtime/pull/63927

Once it's gets merged, I am going to sync this PR with main branch.

Reply:
@stephentoub I've merged the tests into this PR and they are all passing. PTAL